rfcs/20200205-standalone-keras-repository.md

```diff
@@ -0,0 +1,351 @@
+    inputs, 
+    knob1=False,
+    knob2=1)
+
+def new_function(inputs, knob1, knob2=1):
+    ...
+
+==================================
+# PR 33333 (1 file updated)
+# tensorflow/python/keras/layers/core.py
+class existing_layer(Layer):
+
+  def call(inputs):
+<<<
+    array_ops.existing_function(inputs)
+     ...
+>>>
+    array_ops.new_function(
+        inputs, 
+        knob1=True,
+        knob2=3)
+```
+
+
+### Performance Implications
+
+There may be some performance implications as we move towards only using
+public TF APIs. We need to maintain a benchmark to ensure that there
+is no performance regression.
+
+### Dependencies
+
+The TensorFlow pip package will auto-install the Keras package, which shouldn't make 
+any difference on the end-user side. Under the hood, Keras will be a different 
+package imported by `tf_core`, like what we do for TF estimator.
+
+### Developer experience Impact
+
+* The local build and test times should be greatly reduced, since compiling TF
+is no longer needed, and Keras is a pure-Python project.
+* Cross-boundary changes will require some extra handling since such changes 
+needs to be split into two or more PRs. Same for rollbacks.
+* Tooling on the GitHub side (for code review, etc.) is not as good as
+Google-internal tools. This may impact the develoment velocity for
+Keras team members at Google.
+
+### Best Practices, Tutorials and Examples
+
+* The new Keras repository should have a new contribution guide about how to 
+setup a local test environment and iterate based on that. A similar one in 
+tf/addons can be used as an example.
+* The existing TF docs needs to be updated to highlight that Keras code now lives
+in a different repository, with a new process for sending PRs, etc.
+* When filing an issue, people might need to consider where to send the issue,
+e.g. is it a Keras issue or an issue caused by TF but surfaced by Keras. The 
+different ownership of the repository will also cause difficulties for 
+transferring the issue.
+
+### User Impact
+
+* No end-user facing change for current TF users; the split would only affect
+developers, e.g. in-flight PRs during the transition period.
+* For current Keras pip package users, they will get the new TF keras package 
+when they update their pip, which should have more features than the current 
+Keras-team version.
+
+## Questions and Discussion Topics
+
+1. Tools for issue tracking: we can't rely on Google-internal bug tracking tool 
+since it's not publicly visible, also if managing GitHub issues across the orgs 
+is hard, we might need to find some alternatives for tracking bugs/features etc.
+2. OSS tests for TPU-related code. Since TPUs are not available during local
+testing, the verification will have to be done when the PR is mirrored to Google's 
+internal systems.
+3. Transition period for moving the Keras code from `tensorflow/tensorflow` to
+`keras-team/keras`. All in-flight PRs / issues will be affected: they need
```

Comment:
>  All in-flight PRs / issues will be affected

Would it be possible to do a progressive migration? Class after class and function after function or would it become too complicated?

Reply:
We want to minimize the period of transition, since during the period, there will be kind of two source of truth, which might confuse user which one to use/contribute. I will update this RFC with more details about the PR/issue transition.

Reply:
Thanks for the answer! I hope it'll work out! Definitly not an easy task in both cases.

Reply:
Thanks for the answer! I hope it'll work out! Definitly not an easy task in both cases.